hail/src/main/scala/is/hail/backend/service/ServiceBackend.scala

```diff
@@ -363,8 +376,45 @@ class ServiceBackend() extends Backend {
   }
 
   def lowerDistributedSort(ctx: ExecuteContext, stage: TableStage, sortFields: IndexedSeq[SortField], relationalLetsAbove: Map[String, IR]): TableStage = {
-    // Use a local sort for the moment to enable larger pipelines to run
-    LowerDistributedSort.localSort(ctx, stage, sortFields, relationalLetsAbove)
+    val region = ctx.r
+    val rowType = stage.rowType
+    val keyType = rowType.typeAfterSelectNames(sortFields.map(_.field))
+    val rowEType = EType.fromTypeAndAnalysis(rowType, TypeWithRequiredness(rowType)).asInstanceOf[EBaseStruct] // FIXME(danking): Probably not kosher
+    val keyEType = EType.fromTypeAndAnalysis(keyType, TypeWithRequiredness(keyType)).asInstanceOf[EBaseStruct] // FIXME(danking): Probably not kosher
+    val shuffleType = TShuffle(sortFields, rowType, rowEType, keyEType)
+    val shuffleClient = new ShuffleClient(shuffleType, ctx)
+    assert(keyType == shuffleClient.codecs.keyType)
+    val keyDecodedPType = shuffleClient.codecs.keyDecodedPType
+    shuffleClient.start()
+    try {
+      val uuid = shuffleClient.uuid
+
+      val successfulPartitionIds = execute(
+        ctx,
+        stage.mapCollect(relationalLetsAbove)(
+          ShuffleWrite(Literal(shuffleType, uuid), _)))
+
+      val partitionBoundsPointers = shuffleClient.partitionBounds(region, stage.numPartitions) // FIXME(danking): number of partitions should be configurable?
```

Comment:
I wonder where the right place to handle this is. It feels like figuring out how many partitions you actually want might happen in a downstream optimization pass so that any repartitions/joins/etc can be pushed up into the read. Are there cases where you'd want different/configurable behavior by default? Should we be computing the number of new partitions based on the number of rows written to the shuffler?

Reply:
These are good questions, but they're questions for @tpoterba . I'm not sure we do anything that resembles SQL "query planning".

Reply:
this is fine for now. We'll change this as we build in more sensible partition handling

Reply:
Sorry, I was kind of thinking out loud here but my actual comment that I forgot to state was: I don't think you need this FIXME here. Seems like you're doing the right thing at this point, which is just picking a hopefully-reasonable number of partitions here and downstream transformations can repartition as necessary.